File: internal/cli/kamal_fetch.go
package cli

import (
	"errors"
	"fmt"
	"strings"

	"github.com/spf13/cobra"

	"github.com/OpenRangeDevs/wedokeys-cli/internal/client"
	"github.com/OpenRangeDevs/wedokeys-cli/internal/config"
	"github.com/OpenRangeDevs/wedokeys-cli/internal/resolve"
)

// KamalOptions configures the kamal-fetch command.
type KamalOptions struct {
	Account      string // ignored; account is derived from the token
	From         string // project/environment (or just project)
	Env          string
	AllowMissing bool
}

// KamalFetch is the Kamal secrets adapter: it resolves the named aliases and
// prints them as NAME=value lines for Kamal to parse.
func (a *App) KamalFetch(opts KamalOptions, aliases []string) error {
	if len(aliases) == 0 {
		return errors.New("Error: at least one secret name is required.")
	}

	cfg, err := a.newConfig(opts.Env)
	if err != nil {
		return err
	}
	project, environment, err := resolveProjectAndEnv(cfg, opts.From)
	if err != nil {
		return wrapErr(err)
	}
	token, err := cfg.RequireToken()
	if err != nil {
		return wrapErr(err)
	}

	result, err := client.New(cfg.APIURL(), token).ResolveByAliases(aliases, project, environment)
	if err != nil {
		return wrapErr(err)
	}

	// Kamal expects every requested secret; missing ones must fail the fetch.
	if err := resolve.Check(a.err(), result, opts.AllowMissing); err != nil {
		return err
	}
	// Kamal parses one NAME=value per line; multi-line or non-string values
	// cannot be expressed that way, so fail hard rather than emit a corrupt line.
	if err := resolve.AssertSingleLine(a.err(), result.Resolved); err != nil {
		return err
	}

	for _, s := range result.Resolved {
		fmt.Fprintf(a.out(), "%s=%s\n", s.Name, s.Value)
	}
	return nil
}

func resolveProjectAndEnv(cfg *config.Config, from string) (project, environment string, err error) {
	if strings.Contains(from, "/") {
		parts := strings.SplitN(from, "/", 2)
		return parts[0], parts[1], nil
	}
	if from != "" {
		environment, err = cfg.RequireEnvironment()
		return from, environment, err
	}
	project, err = cfg.RequireProjectSlug()
	if err != nil {
		return "", "", err
	}
	environment, err = cfg.RequireEnvironment()
	if err != nil {
		return "", "", err
	}
	return project, environment, nil
}

func newKamalFetchCmd(app *App) *cobra.Command {
	var opts KamalOptions
	cmd := &cobra.Command{
		Use:    "kamal-fetch ALIAS...",
		Short:  "Kamal secrets adapter — fetch secrets by alias (internal use)",
		Hidden: true,
		RunE: func(_ *cobra.Command, args []string) error {
			return app.KamalFetch(opts, args)
		},
	}
	cmd.Flags().StringVar(&opts.Account, "account", "", "Ignored; account is derived from token")
	cmd.Flags().StringVar(&opts.From, "from", "", "project/environment (falls back to wdk.yml + env inference)")
	cmd.Flags().StringVarP(&opts.Env, "env", "e", "", "Environment override")
	cmd.Flags().BoolVar(&opts.AllowMissing, "allow-missing", false, "Proceed even if some aliases fail to resolve")
	return cmd
}

File: ruby-legacy/lib/wdk/commands/kamal_fetch.rb
require_relative "../config"
require_relative "../client"
require_relative "../resolve_guard"

module Wdk
  module Commands
    class KamalFetch
      def initialize(options, alias_names, home_dir: Dir.home, start_dir: Dir.pwd)
        @options     = options
        @alias_names = alias_names
        @home_dir    = home_dir
        @start_dir   = start_dir
      end

      def run
        if @alias_names.empty?
          $stderr.puts "Error: at least one secret name is required."
          exit 1
        end

        config  = Config.new(home_dir: @home_dir, start_dir: @start_dir, env_option: @options["env"])
        project, environment = resolve_project_and_env(config)
        client  = Client.new(api_url: config.api_url, token: config.token!)

        result = client.resolve_by_aliases(
          aliases:     @alias_names,
          project:     project,
          environment: environment
        )

        # Kamal expects every requested secret; missing ones must fail the
        # fetch, not surface later as absent env vars in a deployed app.
        ResolveGuard.check!(result, allow_missing: @options["allow_missing"])

        # Kamal parses our output as one NAME=value per line. A value that spans
        # multiple lines (PEM key, multiline string) or a whole-secret hash (a
        # "*" reference) can't be expressed that way; escaping a newline would
        # corrupt the secret, so we fail hard with a clear message instead.
        assert_single_line_values!(result[:resolved])

        result[:resolved].each do |name, value|
          puts "#{name}=#{value}"
        end
      rescue Client::AuthError => e
        $stderr.puts "Authentication error: #{e.message}"
        exit 1
      rescue Client::NetworkError => e
        $stderr.puts "Network error: #{e.message}"
        exit 1
      rescue Client::ApiError => e
        $stderr.puts "API error: #{e.message}"
        exit 1
      rescue Config::NotConfiguredError, Config::MissingProjectError, Config::MissingEnvironmentError => e
        $stderr.puts "Configuration error: #{e.message}"
        exit 1
      end

      private

      def assert_single_line_values!(resolved)
        offenders = resolved.reject { |_name, value| value.is_a?(String) && !value.match?(/[\r\n]/) }
        return if offenders.empty?

        offenders.each_key do |name|
          $stderr.puts "#{name}: value is not a single line (multiline secret or whole-secret \"*\" reference); " \
                       "reference a specific field instead."
        end
        abort "Error: #{offenders.size} secret(s) cannot be emitted as NAME=value lines for Kamal."
      end

      def resolve_project_and_env(config)
        from = @options["from"]
        if from&.include?("/")
          project, env = from.split("/", 2)
          [ project, env ]
        elsif from
          [ from, config.environment! ]
        else
          [ config.project_slug!, config.environment! ]
        end
      end
    end
  end
end
